src/usr/local/buildpack/utils/folders.sh

```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
+
+# Constants
+ROOT_DIR=/opt/buildpack
+ROOT_UMASK=750
+USER_UMASK=770
+
+# Will set up the general folder structure for the whole buildpack installation
+# This will take $TEST_ROOT into account, so can also be used during testing
+function setup_directories () {
+  local install_dir
+  install_dir=$(get_install_dir)
+
+  mkdir -p "${install_dir}"
+  # contains the installed tools
+  mkdir -m 770 "$(get_tools_path)"
+  # contains env for the installed tools
+  mkdir -m 770 "$(get_env_path)"
+  # contains the latest version of the tools
+  mkdir -m 770 "$(get_version_path)"
+  # contains the wrapper and symlinks for the tools
+  mkdir -m 770 "$(get_bin_path)"
+}
+
+# Creates the given folder path with root and user umask depending on the caller
+# Will also create intermediate folders with correct umask
+function create_folder () {
+  local folder=${1}
+  check folder
+
+  local parent
+  parent=$(dirname "${folder}")
+
+  if [ -d "${folder}" ]; then
+    return
+  fi
+
+  if [ ! -d "${parent}" ]; then
+    create_folder "$parent"
+  fi
+
+  local umask=$USER_UMASK
+  if [ "$(is_root)" -eq 0 ]; then
+    umask=$ROOT_UMASK
+  fi
+
+  mkdir -m "${umask}" "${folder}"
```

Comment:
🤔 this way the user can't install a new node version when one was already installed as root?

I think we need to be able to do that.

Reply:
Correct, we only want to protect root installed versions, not that the user cannot install anything. will change it.

Reply:
Haven't changed this yet, we need to keep this in mind when we are migrating the tools and writing tests for them